polars/polars-ops/src/series/ops/rle.rs

```diff
@@ -29,8 +30,11 @@ pub fn rle_id(s: &Series) -> PolarsResult<Series> {
     let mut out = Vec::with_capacity(s.len());
     out.push(0); // Run numbers start at zero
     s_neq
-        .into_iter()
-        .enumerate()
-        .for_each(|(i, v)| out.push(out[i] + v.unwrap() as u32));
+        .downcast_iter()
+        .for_each(|a| out.extend(a.values_iter().map(|v| v as u32)));
+    out.iter_mut().fold(0, |a, x| {
```

Comment:
Almost there. :)

I think we can do this in a single pass. Something like this


```rust
 
  let mut last_value = 0u32;
   s_neq
        .downcast_iter()
        .for_each(|a| {
            let iter = a.values_iter();

           for v in iter {
              let v = v as u32;
              out.push(last_value + v );
             last_value = v;
           }
});

```

Reply:
And I thought I was into micro-optimizations :-)
Ideally we could do everything, even the comparisons, in a single pass and only store what we need but I didn't see a clear way to do that.

Reply:
> Ideally we could do everything, even the comparisons, in a single pass and only store what we need but I didn't see a clear way to do that.

Yes, that would require us to go down into the known types with some generics. We could follow up with that. The benefit of this implementation is that it has little compiler bloat.

> And I thought I was into micro-optimizations :-)

Haha, I have put a lot of time in making ensuring what we advoncate. A fast dataframe library. I cannot unsee the potential branches, cache misses and allocations ^^